File: scripts/package_extensions.py
#!/usr/bin/env python3
from __future__ import annotations

import json
import shutil
import zipfile
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
SOURCE = ROOT / "extension"
DIST = ROOT / "dist"
FILES = ["background.js", "content.js", "popup.html", "popup.css", "popup.js"]


def build(name: str, manifest: str) -> Path:
    target = DIST / name
    if target.exists():
        shutil.rmtree(target)
    target.mkdir(parents=True)
    for path in SOURCE.iterdir():
        if path.is_file() and path.name not in ("manifest.json", "manifest.firefox.json"):
            shutil.copy2(path, target / path.name)
    manifest_data = json.loads((SOURCE / manifest).read_text(encoding="utf-8"))
    (target / "manifest.json").write_text(json.dumps(manifest_data, indent=2) + "\n", encoding="utf-8")
    legacy_archive = DIST / f"privacy-vision-sih-{name}.zip"
    if legacy_archive.exists():
        legacy_archive.unlink()
    archive = DIST / f"privvy-{name}.zip"
    if archive.exists():
        archive.unlink()
    with zipfile.ZipFile(archive, "w", zipfile.ZIP_DEFLATED) as bundle:
        for path in sorted(target.iterdir()):
            bundle.write(path, path.name)
    return archive


if __name__ == "__main__":
    DIST.mkdir(exist_ok=True)
    for browser, manifest in (("chrome", "manifest.json"), ("firefox", "manifest.firefox.json")):
        print(build(browser, manifest))

File: scripts/package_extensions.mjs
import { mkdir, rm, copyFile, readFile, writeFile, readdir } from 'node:fs/promises';
import { join } from 'node:path';
import { fileURLToPath } from 'node:url';

const root = join(fileURLToPath(new URL('.', import.meta.url)), '..');
const source = join(root, 'extension');
const dist = join(root, 'dist');

const crcTable = new Uint32Array(256);
for (let index = 0; index < 256; index += 1) {
  let value = index;
  for (let bit = 0; bit < 8; bit += 1) value = value & 1 ? 0xedb88320 ^ (value >>> 1) : value >>> 1;
  crcTable[index] = value >>> 0;
}

function crc32(buffer) {
  let crc = 0xffffffff;
  for (const byte of buffer) crc = crcTable[(crc ^ byte) & 0xff] ^ (crc >>> 8);
  return (crc ^ 0xffffffff) >>> 0;
}

function dosTime(date = new Date()) {
  return ((date.getHours() & 0x1f) << 11) | ((date.getMinutes() & 0x3f) << 5) | (Math.floor(date.getSeconds() / 2) & 0x1f);
}

function dosDate(date = new Date()) {
  return (((date.getFullYear() - 1980) & 0x7f) << 9) | (((date.getMonth() + 1) & 0x0f) << 5) | (date.getDate() & 0x1f);
}

function u16(value) {
  const buffer = Buffer.alloc(2);
  buffer.writeUInt16LE(value);
  return buffer;
}

function u32(value) {
  const buffer = Buffer.alloc(4);
  buffer.writeUInt32LE(value >>> 0);
  return buffer;
}

async function zipDirectory(target, archive) {
  const locals = [];
  const centrals = [];
  let offset = 0;
  for (const file of (await readdir(target)).sort()) {
    const name = Buffer.from(file, 'utf8');
    const data = await readFile(join(target, file));
    const crc = crc32(data);
    const local = Buffer.concat([
      u32(0x04034b50), u16(20), u16(0x0800), u16(0), u16(dosTime()), u16(dosDate()),
      u32(crc), u32(data.length), u32(data.length), u16(name.length), u16(0), name, data
    ]);
    const central = Buffer.concat([
      u32(0x02014b50), u16(20), u16(20), u16(0x0800), u16(0), u16(dosTime()), u16(dosDate()),
      u32(crc), u32(data.length), u32(data.length), u16(name.length), u16(0), u16(0), u16(0), u16(0),
      u32(0), u32(offset), name
    ]);
    locals.push(local);
    centrals.push(central);
    offset += local.length;
  }
  const centralDirectory = Buffer.concat(centrals);
  const end = Buffer.concat([
    u32(0x06054b50), u16(0), u16(0), u16(centrals.length), u16(centrals.length),
    u32(centralDirectory.length), u32(offset), u16(0)
  ]);
  await writeFile(archive, Buffer.concat([...locals, centralDirectory, end]));
}

async function build(name, manifestName) {
  const target = join(dist, name);
  await rm(target, { recursive: true, force: true });
  await mkdir(target, { recursive: true });
  for (const file of await readdir(source)) {
    if (file !== 'manifest.json' && file !== 'manifest.firefox.json') {
      await copyFile(join(source, file), join(target, file));
    }
  }
  const manifest = JSON.parse(await readFile(join(source, manifestName), 'utf8'));
  await writeFile(join(target, 'manifest.json'), `${JSON.stringify(manifest, null, 2)}\n`);
  const legacyArchive = join(dist, `privacy-vision-sih-${name}.zip`);
  const archive = join(dist, `privvy-${name}.zip`);
  await rm(legacyArchive, { force: true });
  await rm(archive, { force: true });
  await zipDirectory(target, archive);
  return archive;
}

await mkdir(dist, { recursive: true });
for (const [browser, manifest] of [['chrome', 'manifest.json'], ['firefox', 'manifest.firefox.json']]) {
  console.log(await build(browser, manifest));
}
